src/ppom-controller.ts

```diff
@@ -345,129 +361,216 @@ export class PPOMController extends BaseControllerV2<
+   *
+   * @returns A promise that resolves to an array of new files to download and save to storage.
    */
-  async #fetchBlob(fileUrl: string): Promise<ArrayBuffer> {
-    const response = await safelyExecute(
-      async () => fetch(fileUrl, { cache: 'no-cache' }),
-      true,
+  async #getNewFilesForAllChains(): Promise<void> {
+    const {
+      chainIdCache,
+      storageMetadata,
+      versionInfo: stateVersionInfo,
+    } = this.state;
+
+    // create a map of chainId and files belonging to that chainId
+    const chainIdsFileInfoList = chainIdCache.map(
+      (chain): { chainId: string; versionInfo: PPOMFileVersion[] } => ({
+        chainId: chain.chainId,
+        versionInfo: stateVersionInfo.filter(
+          (versionInfo) =>
+            versionInfo.chainId === chain.chainId &&
+            !this.#checkFilePresentInStorage(storageMetadata, versionInfo),
+        ),
+      }),
     );
 
-    switch (response?.status) {
-      case 200: {
-        return await response.arrayBuffer();
-      }
+    // clear already scheduled fetch if any
+    if (this.#fileScheduleInterval) {
+      clearInterval(this.#fileScheduleInterval);
+    }
 
-      default: {
-        throw new Error(`Failed to fetch file with url ${fileUrl}`);
+    // build a list of files to be fetched for all networks
+    const fileToBeFetchedList: {
+      fileVersionInfo: PPOMFileVersion;
+      isLastFileOfNetwork: boolean;
+    }[] = [];
+    chainIdsFileInfoList.forEach((chainIdFileInfo) => {
+      const { chainId, versionInfo } = chainIdFileInfo;
+      versionInfo.forEach((fileVersionInfo, index) => {
+        fileToBeFetchedList.push({
+          fileVersionInfo,
+          isLastFileOfNetwork: index === versionInfo.length - 1,
+        });
+      });
+      if (versionInfo.length === 0) {
+        // set dataFetched to true for chainId
+        this.#setChainIdDataFetched(chainId);
       }
+    });
+
+    // if schedule interval is large so that not all files can be fetched in
+    // refreshInterval, reduce schedule interval
+    let scheduleInterval = this.state.fileScheduleInterval;
+    if (
+      this.state.refreshInterval / (fileToBeFetchedList.length + 1) <
+      scheduleInterval
+    ) {
+      scheduleInterval =
+        this.state.refreshInterval / (fileToBeFetchedList.length + 1);
     }
+
+    // schedule files to be fetched in intervals
+    this.#fileScheduleInterval = setInterval(() => {
+      const fileToBeFetched = fileToBeFetchedList.pop();
+      if (!fileToBeFetched) {
+        return;
+      }
+      const { fileVersionInfo, isLastFileOfNetwork } = fileToBeFetched;
+      // get the file from CDN
+      this.#getFile(fileVersionInfo)
+        .then(() => {
+          if (isLastFileOfNetwork) {
+            // set dataFetched for chainId to true
+            this.#setChainIdDataFetched(fileVersionInfo.chainId);
+          }
+        })
+        .catch((exp: Error) =>
```

Comment:
We don't have this in `getNewFilesForCurrentChain`, would it help there to log any files that failed to download?

Reply:
When getting files for current network a transaction validation call is received from middleware which logs any error.

But yep will be good to add log here also, let me do that.